### PR TITLE
[FIX] hr_holidays : multiple fixes on accrual_plan

### PR DESCRIPTION
**Pull request by fdamhaut**
See commit msgs

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by fdamhaut** [comment]
@kitan191

File: addons/hr_holidays/models/hr_leave_accrual_plan_level.py [hunk 1]
```diff
@@ -16,14 +16,8 @@
...
-
-def _get_selection_days(self):
-    lang = get_lang(self.env).code
-    return [
-        (DAY_SELECT_VALUES[i - 1],
-        num2words(i, lang=lang, to='ordinal_num') if i < 29 else _('last day'))
-        for i in range(1, 30)
-    ]
+DAY_SELECT_DISPLAY = [str(i) for i in range(1, 29)] + ['last day']
+DAY_SELECT_SELECTION = list(zip(DAY_SELECT_VALUES, DAY_SELECT_DISPLAY))
```
Comment (kitan191) [inline addons/hr_holidays/models/hr_leave_accrual_plan_level.py:20]: If the purpose of the fix is to remove `num2words`, you can just do this to keep it simpler:

```suggestion
def _get_selection_days(self):
    return [
        (DAY_SELECT_VALUES[i - 1],
        str(i) if i < 29 else _('last day'))
        for i in range(1, 30)
    ]
```

You can also remove the import of `num2words`

File: addons/hr_holidays/models/hr_leave_accrual_plan_level.py [hunk 2]
```diff
@@ -74,13 +68,13 @@ class AccrualPlanLevel(models.Model):
     ], default='mon', required=True, string="Allocation on")
     first_day = fields.Integer(default=1)
     first_day_display = fields.Selection(
-        _get_selection_days, compute='_compute_days_display', inverse='_inverse_first_day_display')
+        DAY_SELECT_SELECTION, compute='_compute_days_display', inverse='_inverse_first_day_display')
```
Comment (kitan191) [inline addons/hr_holidays/models/hr_leave_accrual_plan_level.py:71]: This is not required with the suggestion.

File: addons/hr_holidays/models/hr_leave_accrual_plan_level.py [hunk 3]
```diff
@@ -168,13 +162,13 @@ def _compute_level(self):
...
-            level.first_day_display = days_select[min(level.first_day - 1, 28)][0]
-            level.second_day_display = days_select[min(level.second_day - 1, 28)][0]
-            level.first_month_day_display = days_select[min(level.first_month_day - 1, 28)][0]
-            level.second_month_day_display = days_select[min(level.second_month_day - 1, 28)][0]
-            level.yearly_day_display = days_select[min(level.yearly_day - 1, 28)][0]
+            level.first_day_display = days_select[min(level.first_day - 1, 28)][1]
+            level.second_day_display = days_select[min(level.second_day - 1, 28)][1]
+            level.first_month_day_display = days_select[min(level.first_month_day - 1, 28)][1]
+            level.second_month_day_display = days_select[min(level.second_month_day - 1, 28)][1]
+            level.yearly_day_display = days_select[min(level.yearly_day - 1, 28)][1]
```
Comment (kitan191) [inline addons/hr_holidays/models/hr_leave_accrual_plan_level.py:171]: This is not correct.
You should always assign the key part (first element of the tuple) of one of the possible values to a Selection field.

File: addons/hr_holidays/models/hr_leave_accrual_plan_level.py [hunk 4]
```diff
@@ -168,13 +162,13 @@ def _compute_level(self):
 
     @api.depends('first_day', 'second_day', 'first_month_day', 'second_month_day', 'yearly_day')
     def _compute_days_display(self):
-        days_select = _get_selection_days(self)
+        days_select = DAY_SELECT_SELECTION
```
Comment (kitan191) [inline addons/hr_holidays/models/hr_leave_accrual_plan_level.py:165]: This is not required with the suggestion.

**Review by kitan191 — CHANGES_REQUESTED** [review]

File: addons/hr_holidays/models/hr_leave_accrual_plan_level.py (hunk 1 above)
Comment (fdamhaut) [reply addons/hr_holidays/models/hr_leave_accrual_plan_level.py:20]: But isn't it the same just faster ?
(I should still change list to tuple to ensure it's immutable)

Comment (fdamhaut) [reply addons/hr_holidays/models/hr_leave_accrual_plan_level.py:20]: and maybe add a function that return if it's really better not to touch the rest

File: addons/hr_holidays/models/hr_leave_accrual_plan_level.py (hunk 3 above)
Comment (fdamhaut) [reply addons/hr_holidays/models/hr_leave_accrual_plan_level.py:171]: :ok_hand: after further reading now i understand why we translate, it's for showing in selection

File: addons/hr_holidays/models/hr_leave_accrual_plan_level.py (hunk 1 above)
Comment (kitan191) [reply addons/hr_holidays/models/hr_leave_accrual_plan_level.py:20]: Yes, I think it's better to keep the function.

**Comment by fdamhaut** [comment]
Hello @kba-odoo 
mind having a look ?
TY

File: addons/hr_holidays/models/hr_leave_accrual_plan.py [hunk 5]
```diff
@@ -32,7 +32,7 @@ def _compute_level_count(self):
         )
         mapped_count = {group['accrual_plan_id'][0]: group['accrual_plan_id_count'] for group in level_read_group}
         for plan in self:
-            plan.level_count = mapped_count[plan.id]
+            plan.level_count = mapped_count[plan.id] if plan.id in mapped_count else 0
```
Comment (kba-odoo) [inline addons/hr_holidays/models/hr_leave_accrual_plan.py:35]:
```suggestion
            plan.level_count = mapped_count.get(plan.id, 0)
```

File: addons/hr_holidays/models/hr_leave_accrual_plan_level.py [hunk 6]
```diff
@@ -5,25 +5,19 @@
...
 DAY_SELECT_VALUES = [str(i) for i in range(1, 29)] + ['last']
+DAY_SELECT_SELECTION_NO_LAST = tuple(zip(DAY_SELECT_VALUES, (str(i) for i in range(1, 29))))
 
 def _get_selection_days(self):
-    lang = get_lang(self.env).code
-    return [
-        (DAY_SELECT_VALUES[i - 1],
-        num2words(i, lang=lang, to='ordinal_num') if i < 29 else _('last day'))
-        for i in range(1, 30)
-    ]
```
Comment (kba-odoo) [inline addons/hr_holidays/models/hr_leave_accrual_plan_level.py:26]: I'd keep `num2words` here but fallback to English if the language is not supported [1]

[1] : https://github.com/savoirfairelinux/num2words#:~:text=except%20NotImplementedError%3A

**Review by kbapt — CHANGES_REQUESTED** [review]

File: addons/hr_holidays/models/hr_leave_accrual_plan_level.py (hunk 6 above)
Comment (fdamhaut) [reply addons/hr_holidays/models/hr_leave_accrual_plan_level.py:26]: After talking with gmf, we agree it's better to use the number bcs making something grammatically correct in all language is too complex, (aka 2nd of July in english, 2 juillet in french) so using number is easier and more universal.
On top of that, in the xml the use of XXX_day and XXX_day_display seem to be used interchangeably (or at least in similar situation)

Comment (kba-odoo) [reply addons/hr_holidays/models/hr_leave_accrual_plan_level.py:26]: Ok for me if GMF agrees then :-)

**Review by kbapt — APPROVED** [review]

**Comment by kba-odoo** [comment]
@robodoo r+

**Comment by kba-odoo** [comment]
@robodoo rebase-ff

**Pull request closed without merging**